proto/structured.go

```diff
@@ -0,0 +1,185 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package proto
+
+import (
+	"fmt"
+	"strings"
+)
+
+func validateName(name, typ string) error {
+	if len(name) == 0 {
+		return fmt.Errorf("empty %s name", typ)
+	}
+	// TODO(pmattis): Do we want to be more restrictive than this?
+	if strings.Contains(name, "/") {
+		return fmt.Errorf("\"%s\" may not contain \"/\"", name)
+	}
+	return nil
+}
+
+// ValidateTableDesc validates that the table descriptor is well formed. Checks
+// include validating the table, column and index names, verifying that column
+// names and index names are unique and verifying that column IDs and index IDs
+// are consistent.
+func ValidateTableDesc(desc TableDescriptor) error {
+	if err := validateName(desc.Name, "table"); err != nil {
+		return err
+	}
+
+	if len(desc.Columns) == 0 {
+		return fmt.Errorf("table must contain at least 1 column")
+	}
+
+	columnNames := map[string]struct{}{}
+	columnIDs := map[uint32]string{}
+	for _, column := range desc.Columns {
+		if err := validateName(column.Name, "column"); err != nil {
+			return err
+		}
+
+		if _, ok := columnNames[column.Name]; ok {
+			return fmt.Errorf("duplicate column name: \"%s\"", column.Name)
+		}
+		columnNames[column.Name] = struct{}{}
+
+		if other, ok := columnIDs[column.ID]; ok {
+			return fmt.Errorf("column \"%s\" duplicate ID of column \"%s\": %d",
+				column.Name, other, column.ID)
+		}
+		columnIDs[column.ID] = column.Name
+
+		if column.ID >= desc.NextColumnID {
+			return fmt.Errorf("column \"%s\" invalid ID (%d) > next column ID (%d)",
+				column.Name, column.ID, desc.NextColumnID)
+		}
+	}
+
+	// TODO(pmattis): Check that the indexes are unique. That is, no 2 indexes
+	// should contain identical sets of columns.
+
+	if len(desc.Indexes) == 0 {
+		return fmt.Errorf("table must contain at least 1 index")
+	}
+
+	indexNames := map[string]struct{}{}
+	indexIDs := map[uint32]string{}
+	for _, index := range desc.Indexes {
+		if err := validateName(index.Name, "index"); err != nil {
+			return err
+		}
+
+		if _, ok := indexNames[index.Name]; ok {
+			return fmt.Errorf("duplicate index name: \"%s\"", index.Name)
+		}
+		indexNames[index.Name] = struct{}{}
+
+		if other, ok := indexIDs[index.ID]; ok {
+			return fmt.Errorf("index \"%s\" duplicate ID of index \"%s\": %d",
+				index.Name, other, index.ID)
+		}
+		indexIDs[index.ID] = index.Name
+
+		if index.ID >= desc.NextIndexID {
+			return fmt.Errorf("index \"%s\" invalid index ID (%d) > next index ID (%d)",
+				index.Name, index.ID, desc.NextIndexID)
+		}
+
```

Comment:
Its fine. Not a big deal and possibly more complex.

On Mon, Jun 15, 2015 at 10:10 AM Peter Mattis <notifications@github.com>
wrote:

> In proto/structured.go
> <https://github.com/cockroachdb/cockroach/pull/1365#discussion_r32423351>:
>
> > +		if _, ok := indexNames[index.Name]; ok {
> > +			return fmt.Errorf("duplicate index name: \"%s\"", index.Name)
> > +		}
> > +		indexNames[index.Name] = struct{}{}
> > +
> > +		if other, ok := indexIDs[index.ID]; ok {
> > +			return fmt.Errorf("index \"%s\" duplicate ID of index \"%s\": %d",
> > +				index.Name, other, index.ID)
> > +		}
> > +		indexIDs[index.ID] = index.Name
> > +
> > +		if index.ID >= desc.NextIndexID {
> > +			return fmt.Errorf("index \"%s\" invalid index ID (%d) > next index ID (%d)",
> > +				index.Name, index.ID, desc.NextIndexID)
> > +		}
> > +
>
> Not sure it is worth the effort. The code as it currently exists is very
> straightforward at the expense of a little duplication. That is the Go way.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1365/files#r32423351>.
>